app/src/main/java/com/m/services/dataSync/TaskController.java

```diff
@@ -0,0 +1,209 @@
+                    public void onFailure(@NotNull ApolloException e) {
+                        reAuthoriseOn403WithApollo(e);
+                    }
+
+                    @Override
+                    public void onCompleted() {
+                        System.out.println("Subscribed to DeleteTask");
+                    }
+
+                    @Override
+                    public void onTerminated() {
+                        System.out.println("DeleteTask subscription terminated");
+                    }
+
+                    @Override
+                    public void onConnected() {
+                        System.out.println("Connected to DeleteTask subscription");
+                    }
+                });
+    }
+
+    public void deleteTask(String buttonId){
+        AllTasksQuery tasksQuery = AllTasksQuery
+                .builder()
+                .build();
+        client.query(tasksQuery);
+
+        DeleteTaskMutation deleteTask = DeleteTaskMutation
+                .builder()
+                .id(buttonId)
+                .build();
+
+        client.mutate(deleteTask)
+                .refetchQueries(tasksQuery)
+                .enqueue(new ApolloCall.Callback<DeleteTaskMutation.Data>() {
+                    @Override
+                    public void onResponse(@NotNull final Response<DeleteTaskMutation.Data> response) {
+                        listener.deleteTask(response.data().deleteTask());
+                    }
+
+                    @Override
+                    public void onFailure(@NotNull ApolloException e) {
+                        reAuthoriseOn403WithApollo(e);
+                    }
+                });
+    }
+
+    public void createTask(String title, String description){
+        CreateTaskMutation createTask = CreateTaskMutation
+                .builder()
+                .taskTitle(title)
+                .taskDescription(description)
+                .build();
+
+        client.mutate(createTask).enqueue(new ApolloCall.Callback<CreateTaskMutation.Data>() {
+            @Override
+            public void onResponse(@NotNull Response<CreateTaskMutation.Data> response) {
+                System.out.println("Success");
+            }
+
+            @Override
+            public void onFailure(@NotNull ApolloException e) {
+                reAuthoriseOn403WithApollo(e);
+            }
+        });
+    }
+
+    private boolean isOnline() {
+        ConnectivityManager connManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+        NetworkInfo networkInfo = null;
+        if (connManager != null) {
+            networkInfo = connManager.getActiveNetworkInfo();
+        }
+
+        return networkInfo != null && networkInfo.isConnectedOrConnecting();
+    }
+
+
+    private void reAuthoriseOn403WithApollo(ApolloException e){
+        if (e.getMessage().equals("HTTP 403 Forbidden")) {
+            loginController.reAuthorise();
+        }
```

Comment:
this only handles 403 errors. If there are other types of errors returned from the server, nothing will happen in the app. One way we can deal with it is to just show the error to the user. To do that, we can add an error handling method in the listener, and invoke it when there are errors. Then in the activity, we can deal with errors (e.g. show a toast notification).

In fact, I think we probably don't need to use the `loginController` in the `TaskController` here at all. The downside is that now the `TaskController` and `LoginController` are tightly coupled. Instead, if we introduce the error listener, we can pass in the error code to the listener, and in the listener's implementation, we can then decide what to do (e.g. if it's a 403, automatically reauthorise, otherwise show a notification message).